### PR TITLE
Topic/2268 hide drafts from reporter

### PR DESCRIPTION
**Pull request by foxlynx**
#2268 

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue
- [ ] note if PR is on top of other PR
- [ ] note if related change in rems-deploy repo
- [ ] consider adding screenshots for ease of review

## API
- [ ] API is documented and shows up in Swagger UI
- [ ] API is backwards compatible or completely new
- [ ] Events are backwards compatible

## Documentation
- [x] update changelog if necessary
- [ ] add or update docstrings for namespaces and functions
- [ ] components are added to guide page
- [ ] documentation _at least_ for config options (i.e. docs folder)
- [ ] ADR for major architectural decisions or experiments

## Different installations
- [ ] new configuration options added to rems-deploy repository
- [ ] instance specific translations (i.e. LBR kielivara)

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## Accessibility
- [ ] all icons have the aria-label attribute
- [ ] all fields have a label
- [ ] errors are linked to fields with aria-describedby
- [ ] contrast is checked
- [ ] conscious decision about where to move focus after an action

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [ ] no critical TODOs left to implement

### CODE REVIEW

File: src/clj/rems/application/model.clj [hunk 1]
```diff
@@ -663,7 +665,10 @@
     (update application :application/attachments #(filterv visible? %))))
 
 (defn see-application? [application user-id]
-  (not= #{:everyone-else} (permissions/user-roles application user-id)))
+  (let [permissions (permissions/user-roles application user-id)]
+    (if (and (= :application.state/draft (:application/state application)) (contains? permissions :reporter))
+      false
+      (not= #{:everyone-else} permissions))))
```
Comment (opqdonut) [inline src/clj/rems/application/model.clj:671]: I feel like this logic could be a bit more clear. Now it disallows reporters from seeing drafts. I'd rather only allow applicants (and members?) to see drafts. So something like

```
(if draft
  (user-is-applicant-or-member)
  (not= #{:everyone-else permissions))

File: test/clj/rems/application/test_model.clj [hunk 2]
```diff
@@ -1432,11 +1432,15 @@
...
+                        (permissions/give-role-to-users :reporter ["reporter"])
                         (permissions/give-role-to-users :role-1 ["user-1"])
                         (permissions/give-role-to-users :role-2 ["user-2"])
                         (permissions/update-role-permissions {:role-1 #{}
-                                                              :role-2 #{:foo :bar}}))
+                                                              :role-2 #{:foo :bar}
+                                                              :reporter #{:see-everything}}))
         enriched (model/enrich-with-injections application injections)]
+    (testing "reporter can't see draft application"
+      (is (nil? (model/apply-user-permissions enriched "reporter"))))
```
Comment (opqdonut) [inline test/clj/rems/application/test_model.clj:1443]: this is a good test, but I'd also like an API-level test

**Review by opqdonut — CHANGES_REQUESTED** [review]

File: test/clj/rems/api/test_audit_log.clj [hunk 3]
```diff
@@ -100,7 +105,7 @@
                        read-ok-body)))))
     (testing "filtering log by user"
       (is (= [{:userid "alice" :apikey "42" :method "get" :path "/api/users/active" :status "403"}
-              {:userid "alice" :apikey "42" :method "get" :path (str "/api/applications/" app-id) :status "200"}
+              {:apikey "42", :method "get", :path (str "/api/applications/" app-id), :status "200", :userid "alice"}
```
Comment (opqdonut) [inline test/clj/rems/api/test_audit_log.clj:108]: the change on this line doesn't really do anythign does it? merely breaks the symmetry with the other lines

File: src/clj/rems/application/model.clj (hunk 1 above)
Comment (opqdonut) [reply src/clj/rems/application/model.clj:671]: I'd still like to see this clarified

**Review by opqdonut — DISMISSED** [review]

File: test/clj/rems/application/test_model.clj (hunk 2 above)
Comment (opqdonut) [reply test/clj/rems/application/test_model.clj:1443]: thanks!

File: test/clj/rems/api/test_audit_log.clj (hunk 3 above)
Comment (foxlynx) [reply test/clj/rems/api/test_audit_log.clj]: Fixed

File: src/clj/rems/application/model.clj [hunk 4]
```diff
@@ -664,11 +664,15 @@
...
+        (contains? roles :member))))
+
 (defn see-application? [application user-id]
-  (let [permissions (permissions/user-roles application user-id)]
-    (if (and (= :application.state/draft (:application/state application)) (contains? permissions :reporter))
-      false
-      (not= #{:everyone-else} permissions))))
+  (if (= :application.state/draft (:application/state application))
+    (user-is-applicant-or-member application user-id)
+    (not= #{:everyone-else})))
```
Comment (opqdonut) [inline src/clj/rems/application/model.clj:675]: missing the other argument to `not=` here :)

Comment (foxlynx) [reply src/clj/rems/application/model.clj:675]: ;D

File: test/clj/rems/api/test_applications.clj [hunk 5]
```diff
@@ -489,6 +489,12 @@
                (authenticate api-key "bob")
                handler))))
 
+    (testing "seeing draft application as reporter is forbidden"
+      (is (response-is-forbidden?
+            (-> (request :get (str "/api/applications/" application-id))
+                (authenticate api-key "reporter")
+                handler))))
+
```
Comment (Macroz) [inline test/clj/rems/api/test_applications.clj:497]: Could combine the two seeing is forbidden? Just reduce the duplication a bit or introduce even more cases?

File: test/clj/rems/application/test_model.clj [hunk 6]
```diff
@@ -1447,40 +1477,41 @@
...
 
       (testing "normal users"
         (let [application (model/apply-user-permissions enriched "user-2")]
           (testing "see only some events"
-            (is (= restricted-event-ids
-                   (mapv :event/id (:application/events application)))))
+            (is (= [{:event/type :application.event/created}
+                    {:event/type :application.event/submitted}
+                    {:event/type :application.event/remarked
+                     :application/comment "this is public"}]
```
Comment (Macroz) [inline test/clj/rems/application/test_model.clj:1513]: For this particular code I would have gone with format like 
```
            (is (= [{:event/type :application.event/created}
                    {:event/type :application.event/submitted}
                    {:event/type :application.event/remarked :application/comment "this is public"}]

```

**Review by Macroz — APPROVED** [review]
Looks good to me!

**Pull request merged**